File: Palindrome_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

public class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode root = head;
        ListNode list1 = reverseLink(root);
        ListNode list2 = root;
    
        while(list1 != null && list2 != null) {
            if(list1.val != list2.val) return false;
            list1 = list1.next;
            list2 = list2.next;
        }
        return list1 == null && list2 == null;
    }
    
    private ListNode reverseLink(ListNode head) {
        ListNode ret = null;
        while(head != null) {
            ListNode tmp = new ListNode(head.val);
            tmp.next = ret;
            ret = tmp;
            head = head.next;
        }
        return ret;
    }
}

File: Swift/palindrome-linked-list.swift
/**
 * https://leetcode.com/problems/palindrome-linked-list/
 * 
 * 
 */ 
// Date: Mon May  4 23:13:56 PDT 2020
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    /// - Complexity:
    ///     - Time: O(n)
    ///     - Space: O(n), space used to reverse the new created ListNode.
    ///
    func isPalindrome(_ head: ListNode?) -> Bool {
        var head1 = head
        var head2 = reverseList(head)
        // printNode(head1)
        // printNode(head2)
        while head1 != nil, head2 != nil {
            if head1?.val != head2?.val { return false }
            head1 = head1?.next
            head2 = head2?.next
        }
        return true
    }
    
    private func reverseList(_ head: ListNode?) -> ListNode? {
        guard let h = head else { return head }
        var next = h.next
        var head = ListNode(h.val)
        while next != nil {
            let nnext = next?.next
            let node = ListNode(next!.val, head)
            head = node
            next = nnext
        }
        return head
    }
    
    private func printNode(_ head: ListNode?) {
        guard let head = head else { return }
        print(" \(head.val) ")
        self.printNode(head.next)
    }
}
